src/UniswapV4MulticurveRehypeInitializerHook.sol

```diff
@@ -0,0 +1,612 @@
+        Position storage position = getPosition[poolId];
+        if (position.salt == bytes32(0)) {
+            position.tickLower = int24(TickMath.minUsableTick(key.tickSpacing));
+            position.tickUpper = int24(TickMath.maxUsableTick(key.tickSpacing));
+            position.salt = _fullRangeSalt(poolId);
+        }
+
+        getFees[poolId] = Fees({ fees0: 0, fees1: 0, customFee: 3000 });
+
+        return BaseHook.afterInitialize.selector;
+    }
+
+    /// @inheritdoc BaseHook
+    function _beforeAddLiquidity(
+        address sender,
+        PoolKey calldata,
+        IPoolManager.ModifyLiquidityParams calldata,
+        bytes calldata
+    ) internal view override returns (bytes4) {
+        if (sender == address(this) || sender == INITIALIZER) {
+            return BaseHook.beforeAddLiquidity.selector;
+        }
+        revert OnlyInitializer();
+    }
+
+    /// @inheritdoc BaseHook
+    function _afterAddLiquidity(
+        address,
+        PoolKey calldata key,
+        IPoolManager.ModifyLiquidityParams calldata params,
+        BalanceDelta,
+        BalanceDelta,
+        bytes calldata
+    ) internal override returns (bytes4, BalanceDelta) {
+        PoolId poolId = key.toId();
+        Position storage storedPosition = getPosition[poolId];
+        int256 liquidityDelta = params.liquidityDelta;
+
+        if (liquidityDelta != 0) {
+            if (storedPosition.salt == bytes32(0)) {
+                int24 minTick = TickMath.minUsableTick(key.tickSpacing);
+                int24 maxTick = TickMath.maxUsableTick(key.tickSpacing);
+                if (params.tickLower == minTick || params.tickUpper == maxTick) {
+                    storedPosition.tickLower = params.tickLower;
+                    storedPosition.tickUpper = params.tickUpper;
+                    storedPosition.salt = params.salt;
+                    if (liquidityDelta > 0) {
+                        storedPosition.liquidity = uint128(uint256(liquidityDelta));
+                    }
+                }
+            } else if (storedPosition.salt == params.salt) {
+                if (liquidityDelta > 0) {
+                    storedPosition.liquidity += uint128(uint256(liquidityDelta));
+                } else {
+                    storedPosition.liquidity -= uint128(uint256(-liquidityDelta));
+                }
+            }
+        }
+
+        emit ModifyLiquidity(key, params);
+        return (BaseHook.afterAddLiquidity.selector, BalanceDeltaLibrary.ZERO_DELTA);
+    }
+
+    /// @inheritdoc BaseHook
+    function _afterRemoveLiquidity(
+        address,
+        PoolKey calldata key,
+        IPoolManager.ModifyLiquidityParams calldata params,
+        BalanceDelta,
+        BalanceDelta,
+        bytes calldata
+    ) internal override returns (bytes4, BalanceDelta) {
+        emit ModifyLiquidity(key, params);
+        return (BaseHook.afterRemoveLiquidity.selector, BalanceDeltaLibrary.ZERO_DELTA);
+    }
```

Comment:
Shouldn't we also be updating `storedPosition.liquidity` when removing liquidity?